### PR TITLE
Ship Typescript types for Matrix APIs in @matrix-org/spec

### PR DESCRIPTION
**Pull request by t3chguy**
Example consumer: https://github.com/matrix-org/matrix-js-sdk/pull/3653

Preview: https://pr1766--matrix-spec-previews.netlify.app

### CODE REVIEW

File: packages/npm/openapi.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+SPEC_DIR="v${RELEASE}"
+if [ "$RELEASE" == "0.0" ]; then
+  RELEASE="unstable"
+  SPEC_DIR="/unstable"
+fi
+
+BASE_URL="https://spec.matrix.org/$SPEC_DIR"
+
+for api in application-service client-server push-gateway server-server; do
+  FILE="$api-api.json"
```
Comment (richvdh) [inline packages/npm/openapi.sh:15]: can we give this a better name? And a comment?

```suggestion
  # Combine the separate OpenAPI definition files into one document
  combined_doc="${api}-api.json"
```

**Review by richvdh — COMMENTED** [review]
I love this idea: really interesting, thanks!

But I'm not entirely convinced it belongs in the main matrix-spec repo. It feels like the thin end of a wedge, where before too long we'll be maintaining bindings for typescript, kotlin, go, rust, and everything else under the sun. My experience suggests that once you go down that path, releases become very fragile (you can bet that at least one of your target platforms will break), and it's also difficult to make incremental releases that fix things in individual target platforms.

Long story short: I feel like this should be in a separate repo, publishing separate artifacts to npm, based on the published OpenAPI documents from the spec, (eg https://spec.matrix.org/v1.10/server-server-api/api.json).

Happy to hear contrary opinions though.

File: packages/npm/openapi.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  ../../scripts/dump-openapi.py \
+    --base-url "$BASE_URL" \
+    --api "$api" \
+    -r "$RELEASE" \
+    -o "$FILE"
+
+  yarn openapi-typescript "$FILE" --output "$api.d.ts"
+  # We remove these lines to workaround dodgy types
+  sed -i.bak "/\[key: string\]: Record<string, never> \| undefined;/d" client-server.d.ts
+  rm "$api.d.ts.bak"
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply packages/npm/openapi.sh:25]: This is not BSD Sed compatible, so would break macOS compatibility

File: packages/npm/openapi.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+for api in application-service client-server push-gateway server-server; do
+  FILE="$api-api.json"
+  ../../scripts/dump-openapi.py \
+    --base-url "$BASE_URL" \
+    --api "$api" \
+    -r "$RELEASE" \
+    -o "$FILE"
+
+  yarn openapi-typescript "$FILE" --output "$api.d.ts"
+  # We remove these lines to workaround dodgy types
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply packages/npm/openapi.sh:23]: The well-known response for example had a `[key: string]: Record<string, never> | undefined;` which conflicted with the other fields on that response which were not of the type `Record<string, never>;`

**Comment by t3chguy** [comment]
> I feel like this should be in a separate repo

Happy to move it but I won't be maintaining multiple npm packages so I'd be taking the current package name with me

File: packages/npm/openapi.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply packages/npm/openapi.sh:25]: ok maybe link to the upstream bug in the comment?

File: packages/npm/openapi.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply packages/npm/openapi.sh:27]: ugh

**Comment by richvdh** [comment]
> Happy to move it but I won't be maintaining multiple npm packages so I'd be taking the current package name with me

Yup, I (and other members of the SCT) think it makes sense to have a single npm package. Presumably you'd have to pull `sas-emoji.json` over HTTP from github during the build process, or something?

Sounds good, anyway. If you'd like to make this happen, it would be great!

**Comment by t3chguy** [comment]
Replaced by https://github.com/t3chguy/matrix-spec-npm
